File: packages/ballerina-central/src/views/overview.ts
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * `bal-library <pkg>` — the default document, and the only one an agent cannot
 * skip.
 *
 * It carries the readme, every client's constructor and functions, the
 * module-level functions, and the error declarations. NO OTHER TYPES: they are 738
 * of `ballerinax/github`'s 927KB, and excluding them is the whole reason one
 * document is viable. Every other type arrives named in a return type or a
 * parameter, which is the ordering that cannot go wrong — github's own readme
 * example returns `Repository` while the operation returns `FullRepository`, and
 * eleven records in the package carry a `stargazersCount`, four of them optional.
 *
 * ERRORS ARE THE ONE TYPE FAMILY THAT STAYS, because they are the only
 * declarations unreachable from a signature. `ballerinax/github` declares zero
 * errors and all 903 of its operations return the language-level `error`, so
 * nothing in the API document names `http:ClientRequestError` — which is exactly
 * the lookup eight of nine recorded runs came for. They are also cheap: 3,089
 * bytes for http's 56, at most 334 for anything else, and the section is omitted
 * entirely for the four packages that declare none.
 */

import { code, columns, count, Report } from "../report.js";
import type { LoadedPackage } from "../library.js";
import type { ClientClass, Fn, Library, TypeDef } from "../model.js";
import { formatQualifiedName } from "../qualified.js";
import { demoteHeadings } from "../readme.js";
import { renderMemberFunction, renderStandaloneFunction } from "../render/signature.js";
import { renderTypeDef } from "../render/typedef.js";
import { buildPathTree, operationsOf, type PathNode } from "../symbols.js";

/**
 * When a client's resource functions are replaced by a path tree.
 *
 * Both halves matter and they are not redundant. The COUNT is the readable limit;
 * the BYTE guard is what stops a verbose 80-operation connector from producing a
 * 40KB document that passes the count. At gmail's measured ~480 bytes per
 * operation, 100 operations is about 48KB, which is far too large for a document
 * meant to be read from the top. On the nine-fixture corpus both halves agree —
 * github and slack take the tree, gmail at 32 operations and 15.3KB stays inline —
 * so the guard costs nothing today and is there for the package that has not been
 * recorded yet.
 */
export const MAX_INLINE_OPERATIONS = 100;
export const MAX_INLINE_SIGNATURE_BYTES = 20_000;

export interface OverviewOptions {
  /** Narrow to one client. Everything else about the document is unchanged. */
  readonly client?: string;
}

function bytesOf(text: string): number {
  return Buffer.byteLength(text, "utf-8");
}

/**
 * How the declarations break down, for the one line that says they are not here.
 *
 * Errors are excluded from the total because they ARE listed, in their own
 * section; counting them under "not listed here" would make the line false by
 * exactly the number of errors.
 */
function describeTypes(typeDefs: readonly TypeDef[]): string {
  const kinds = typeDefs.filter((typeDef) => typeDef.kind !== "error");
  if (kinds.length === 0) return "none declared";
  const records = kinds.filter((typeDef) => typeDef.kind === "record").length;
  const unions = kinds.filter((typeDef) => typeDef.kind === "union").length;
  const other = kinds.length - records - unions;
  const parts = [`${count(records)} records`, `${count(unions)} unions`, `${count(other)} other`].filter(
    (part) => !part.startsWith("0 "),
  );
  return `${count(kinds.length)} declarations (${parts.join(", ")}), not listed here — read one with \`type\``;
}

function errorsOf(library: Library): readonly Extract<TypeDef, { kind: "error" }>[] {
  return library.typeDefs.flatMap((typeDef) => (typeDef.kind === "error" ? [typeDef] : []));
}

/** A signature at column zero — a report quotes declarations, it does not indent them. */
function signature(fn: Fn): string {
  return renderMemberFunction(fn, "");
}

/**
 * One client's section.
 *
 * Functions are GROUPED BY CALL SYNTAX, and the heading states it rather than
 * leaving it to be inferred from a keyword. ADR-0001 already had to fix one
 * rendering bug of exactly this kind — upstream declared `remote function close()`
 * on a database client, and `dbClient->close()` does not compile — so "remote
 * means `->`, normal means `.`" is a fact worth writing down where the signature
 * is read.
 */
function clientSection(report: Report, client: ClientClass, pkg: string): void {
  report.heading(2, `Client ${code(client.name)}`);
  if (client.description.trim() !== "") report.paragraph(client.description.split("\n")[0] ?? "");

  const constructor = client.functions.find((fn) => fn.kind === "constructor");
  if (constructor) {
    report.heading(3, "Constructor");
    report.ballerina([signature(constructor)]);
  }

  const remote = client.functions.filter((fn) => fn.kind === "remote");
  if (remote.length > 0) {
    report.heading(3, `Remote functions — ${count(remote.length)}, call with ${code("->")}`);
    report.ballerina(remote.map(signature));
  }

  const normal = client.functions.filter((fn) => fn.kind === "normal");
  if (normal.length > 0) {
    report.heading(3, `Normal functions — ${count(normal.length)}, call with ${code(".")}`);
    report.ballerina(normal.map(signature));
  }

  resourceSection(report, client, pkg);
}

function resourceSection(report: Report, client: ClientClass, pkg: string): void {
  const operations = operationsOf(client);
  if (operations.length === 0) return;

  const rendered = operations.map((operation) => signature(operation.fn));
  const bytes = rendered.reduce((sum, line) => sum + bytesOf(line) + 1, 0);
  const tooMany = operations.length > MAX_INLINE_OPERATIONS;
  const tooBig = bytes > MAX_INLINE_SIGNATURE_BYTES;

  report.heading(3, `Resource functions — ${count(operations.length)}, call with ${code("->")} and a path`);

  if (!tooMany && !tooBig) {
    report.ballerina(rendered);
    return;
  }

  report.paragraph(
    `**Not listed here** — ${count(operations.length)} operations, ${count(bytes)} bytes of signatures, ` +
      `over the ${tooMany ? `${count(MAX_INLINE_OPERATIONS)}-operation` : `${count(MAX_INLINE_SIGNATURE_BYTES)}-byte`} limit. ` +
      `Top-level path segments, with the number of operations under each:`,
  );

  const tree = buildPathTree(operations);
  report.literal(columns(tree.children.map((child) => `${child.segment} ${count(child.total)}`)));
  report.paragraph(describeTree(tree));
  report.bullets([
    `navigate: ${code(`bal-library ops ${pkg} <segment>`)}`,
    `signatures under a path: ${code(`bal-library ops ${pkg} '<path>' --sigs`)}`,
  ]);
}

/**
 * An honest header for a "tree" that is not always one.
 *
 * `ballerinax/slack` has 174 operations across 174 distinct top-level segments —
 * its paths are RPC-style single names like `chat.postMessage` — so level 1 IS the
 * complete operation list. That needs no special case, only saying so, because an
 * agent told to descend a flat list will spend a turn discovering there is nowhere
 * to descend to.
 */
function describeTree(tree: PathNode): string {
  const flat = tree.children.every((child) => child.children.length === 0);
  return flat
    ? `${count(tree.total)} operations across ${count(tree.children.length)} top-level segments, and none of them nests — the list above is every operation. Nothing to descend into.`
    : `${count(tree.total)} operations across ${count(tree.children.length)} top-level segments.`;
}

export function renderOverview(loaded: LoadedPackage, options: OverviewOptions = {}): string {
  const pkg = formatQualifiedName(loaded.qualified);
  const { library } = loaded;
  const clients =
    options.client === undefined
      ? library.clients
      : library.clients.filter((client) => client.name === options.client);

  const report = new Report("overview");
  report.heading(1, `${pkg} ${loaded.version}`);

  const errors = errorsOf(library);
  report.facts([
    ["Source", loaded.provenance],
    [
      "Clients",
      library.clients.length === 0
        ? "none — the callable surface is module-level functions"
        : library.clients.map((client) => code(client.name)).join(", "),
    ],
    [
      "Module functions",
      library.functions.length === 0 ? "none" : `${count(library.functions.length)}, listed below`,
    ],
    [
      "Errors",
      errors.length === 0
        ? "none declared; operations return the language-level `error`"
        : `${count(errors.length)}, listed below`,
    ],
    ["Types", describeTypes(library.typeDefs)],
  ]);

  if (options.client !== undefined && clients.length === 0) {
    report.paragraph(
      `**No client named ${code(options.client)}.** The full document is ${code(`bal-library ${pkg}`)}.`,
    );
    return report.toString();
  }

  report.heading(2, "Next");
  report.bullets([
    `${code(`bal-library ops ${pkg} <path>`)} — navigate a client's operations`,
    `${code(`bal-library type ${pkg} <Name> [--deps]`)} — read a declaration whole`,
    `${code(`bal-library api ${pkg}`)} — every declaration, when nothing above answered`,
  ]);

  for (const client of clients) clientSection(report, client, pkg);

  if (library.functions.length > 0) {
    report.heading(2, `Module-level functions — ${count(library.functions.length)}, call with ${code(".")}`);
    report.ballerina(library.functions.map(renderStandaloneFunction));
  }

  if (errors.length > 0) {
    report.heading(2, `Errors — ${count(errors.length)}`);
    report.paragraph(
      `The subtype chain is what ${code("is")} tests against, so ${code("e is <Name>")} works off these lines directly.`,
    );
    report.ballerina(errors.map(renderTypeDef));
  }

  guideSection(report, loaded);
  return report.toString();
}

/**
 * The package's own guide, last, with its headings demoted two levels.
 *
 * Last because it is the largest section for most packages and the document is
 * meant to be read from the top with the API first. Demoted so the document keeps
 * ONE outline and `grep '^## '` returns the overview's sections rather than the
 * readme's. Otherwise verbatim, fences intact and copyable — it goes in as raw
 * Markdown rather than as commented-out text, which reads better and costs three
 * fewer bytes per line.
 */
function guideSection(report: Report, loaded: LoadedPackage): void {
  if (loaded.readmes.length === 0) {
    report.heading(2, "Guide");
    report.paragraph("This package publishes no guide. The signatures above are all Central holds.");
    return;
  }

  for (const readme of loaded.readmes) {
    report.heading(2, loaded.readmes.length === 1 ? "Guide" : `Guide — ${readme.module}`);
    report.paragraph("*The package's own readme, verbatim, with its headings demoted two levels.*");
    report.embedded(demoteHeadings(readme.markdown, 2));
  }
}

File: packages/bal-library-tool/native/src/main/java/io/ballerina/library/views/Overview.java
/*
 *  Copyright (c) 2026, WSO2 LLC. (http://www.wso2.com)
 *
 *  WSO2 LLC. licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except
 *  in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

package io.ballerina.library.views;

import io.ballerina.library.LoadedPackage;
import io.ballerina.library.Texts;
import io.ballerina.library.model.Fn;
import io.ballerina.library.model.Library;
import io.ballerina.library.model.TypeDef;
import io.ballerina.library.render.Report;
import io.ballerina.library.symbols.Filter;
import io.ballerina.library.symbols.Surface;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * {@code overview <pkg> [-s <q>]} — a bounded MAP of the package, and the entry document.
 *
 * <p><b>NO GENERATED SIGNATURES.</b> This is the largest behavioural change in the redesign and it is chosen
 * deliberately over a byte cap. A cap still lets {@code ballerina/crypto} emit 20,000 bytes before it degrades; a
 * map is bounded BY CONSTRUCTION, so crypto goes from 1,177 lines to roughly twenty and the failure where an eval
 * harness silently substituted a 2.2KB stub disappears rather than becoming rarer.
 *
 * <p>The cost is honest and worth naming: {@code googleapis.sheets} goes from one call to two, because its
 * 208-line overview carried all 43 remote signatures. That is why the 20,000-byte signature budget moved onto the
 * container verbs — the second call returns all 43 in full.
 *
 * <p>What remains is what a map is for: the counts, the readme's quickstart, the guide's chunk index, where to go
 * next, and a roster whose every row ends in the command that opens it. A roster row that dead-ends is the one
 * shape this document never ships (ADR-0019) — {@code overview}'s old unconditional {@code ops <pkg> <path>}
 * pointer answered "none in any client" on {@code ballerinax/aws.s3} and pointed back here, a two-call loop
 * carrying no information.
 *
 * <p><b>ORDERED TO SURVIVE A PIPE</b> (ADR-0017, unchanged and strengthened). 80% of recorded lookups are piped
 * through a filter, and putting {@code ## Next} before the bulk took {@code head -100} reaching it from 0 of 11
 * packages to 11 of 11.
 *
 * <p>{@code -s} makes this the CROSS-KIND search — the "which symbol owns this capability?" question that no
 * kind-specific verb can answer, since the caller does not yet know the kind.
 *
 * @since 0.1.0
 */
public final class Overview {

    /**
     * How many roster rows the map prints before it prints a count instead.
     *
     * <p>{@code ballerina/http} declares 91 classes and {@code ballerinax/postgresql} 126. A roster is read to
     * CHOOSE one, and a list nobody can hold is not a choice — the search line beside it is.
     */
    public static final int MAX_ROSTER_ROWS = 20;

    /** How many results a cross-kind search lists before it names the count and the narrowing. */
    public static final int MAX_SEARCH_ROWS = 30;

    /** The one qualifier a type alias may carry only if it is an error type. */
    private static final String DISTINCT = "distinct ";

    /** How much of a facts row the error names may take before the command replaces them. */
    private static final int MAX_ERROR_NAMES_CHARS = 300;

    private Overview() {
    }

    /** @param search the {@code -s} query, or {@code null} for the map */
    public record Options(String search) {

        public static final Options MAP = new Options(null);

        public boolean filtered() {
            return search != null && !search.isBlank();
        }
    }

    public static String render(LoadedPackage loaded) {
        return render(loaded, Options.MAP);
    }

    public static String render(LoadedPackage loaded, Options options) {
        return options.filtered() ? searched(loaded, options) : map(loaded);
    }

    // -----------------------------------------------------------------------
    // The map
    // -----------------------------------------------------------------------

    private static String map(LoadedPackage loaded) {
        String pkg = loaded.qualified().qualified();
        Library library = loaded.library();

        Report report = new Report("overview");
        report.heading(1, pkg + " " + loaded.version().text());
        report.facts(factsOf(loaded, library, pkg));

        chunkIndex(report, loaded, pkg);

        report.heading(2, "Next");
        report.bullets(nextBullets(library, pkg));

        roster(report, "Clients", Surface.Scope.CLIENT, library, pkg);
        roster(report, "Classes and object types", Surface.Scope.CLASS, library, pkg);
        roster(report, "Module-level functions", Surface.Scope.MODULE, library, pkg);
        // LAST, and that is the whole reason the section can be unbounded. ADR-0024 took the cap off the quoted
        // code, so this is the one part of the map whose length the package decides; everything a reader
        // navigates by — the facts, the chunk index, `## Next`, the rosters — is already behind them by here, so
        // a `head -100` loses trailing examples rather than the map.
        quickstart(report, loaded, pkg);
        return report.toString();
    }

    private static List<Report.Fact> factsOf(LoadedPackage loaded, Library library, String pkg) {
        List<Report.Fact> facts = new ArrayList<>(Report.warning(loaded.warning()));
        facts.add(new Report.Fact("Clients", describeScope(library, Surface.Scope.CLIENT, pkg)));
        facts.add(new Report.Fact("Classes", describeScope(library, Surface.Scope.CLASS, pkg)));
        facts.add(new Report.Fact("Module functions", describeScope(library, Surface.Scope.MODULE, pkg)));
        facts.add(new Report.Fact("Errors", describeErrors(errorsOf(library), pkg)));
        facts.add(new Report.Fact("Types", describeTypes(library.declarations())));
        facts.add(new Report.Fact("Guide", describeGuide(loaded, pkg)));
        return facts;
    }

    /**
     * One scope, counted and pointed at.
     *
     * <p>Read off {@link Surface} rather than off Central's {@code clients} array, which is not the callable
     * surface: SQL-03 recorded the entry document asserting "none — the callable surface is module-level
     * functions" for {@code ballerina/sql}, which has two clients and whose own readme quotes
     * {@code dbClient->query(query)} a few lines further down. An assertion ends a search where an omission would
     * only have invited a follow-up.
     */
    private static String describeScope(Library library, Surface.Scope scope, String pkg) {
        List<Surface.Container> containers = Surface.of(library, scope);
        if (containers.isEmpty()) {
            return "none";
        }
        if (scope == Surface.Scope.MODULE) {
            return Texts.count(containers.get(0).functions().size()) + " — "
                    + Texts.code("bal library funcs " + pkg);
        }
        String named = containers.stream().map(Surface.Container::name).map(Texts::code)
                .collect(Collectors.joining(", "));
        String command = Texts.code("bal library " + scope.verb() + " " + pkg);
        return named.length() <= MAX_ERROR_NAMES_CHARS
                ? Texts.count(containers.size()) + " — " + named + " — " + command
                : Texts.count(containers.size()) + ", too many to name here — " + command;
    }

    /**
     * The error declarations, NAMED on the row rather than listed in a section.
     *
     * <p>The names could not leave with the section: {@code BucketAlreadyOwnedByYouError} and
     * {@code InvalidRangeError} are not guessable, and a row reading "6, listed below" with nothing below names a
     * fact and withholds the one token needed to reach it. {@code ballerina/http}'s 65 are the only case in the
     * corpus that does not fit a line.
     */
    private static String describeErrors(List<TypeDef> errors, String pkg) {
        if (errors.isEmpty()) {
            return "none declared here; each operation names its error type in its `returns` clause";
        }
        String read = "read one with " + Texts.code("bal library type " + pkg + " <Name>");
        String named = errors.stream().map(TypeDef::name).map(Texts::code)
                .collect(Collectors.joining(", "));
        return named.length() <= MAX_ERROR_NAMES_CHARS
                ? Texts.count(errors.size()) + " — " + named + " — " + read
                : Texts.count(errors.size()) + ", too many to name here — " + read;
    }

    private static String describeGuide(LoadedPackage loaded, String pkg) {
        int lines = loaded.readmes().stream()
                .mapToInt(readme -> readme.markdown().split("\n", -1).length)
                .sum();
        return lines == 0
                ? "none published"
                : Texts.count(lines) + " lines — " + Texts.code("bal library guide " + pkg);
    }

    /**
     * How the declarations break down, for the one line that says they are not here.
     *
     * <p>Errors are excluded from the total because they ARE named, on their own row; counting them under "not
     * listed here" would make the line false by exactly the number of errors. SHEETS-04 and GMAIL-05: enums,
     * constants and objects each get their own count — folded into "other" they were invisible, so the entry
     * document never said a package HAS enums, and an enum is the one declaration whose members cannot be
     * guessed.
     */
    private static String describeTypes(List<TypeDef> typeDefs) {
        List<TypeDef> kinds = typeDefs.stream()
                .filter(typeDef -> !isErrorDeclaration(typeDef))
                .toList();
        if (kinds.isEmpty()) {
            return "none declared";
        }
        Map<String, Long> counts = new LinkedHashMap<>();
        counts.put("records", kinds.stream().filter(TypeDef.Rec.class::isInstance).count());
        // "type aliases" and not "unions": the count covers all seventeen of Central's alias categories, and a
        // union is one of them.
        counts.put("type aliases", kinds.stream().filter(TypeDef.Alias.class::isInstance).count());
        counts.put("enums", kinds.stream().filter(TypeDef.Enumeration.class::isInstance).count());
        counts.put("constants", kinds.stream().filter(TypeDef.Constant.class::isInstance).count());
        counts.put("classes and object types", kinds.stream().filter(TypeDef.ObjectDef.class::isInstance)
                .count());
        counts.put("module-level variables", kinds.stream().filter(TypeDef.Variable.class::isInstance).count());
        long other = kinds.size() - counts.values().stream().mapToLong(Long::longValue).sum();
        counts.put("other", other);

        List<String> parts = new ArrayList<>();
        counts.forEach((label, count) -> {
            if (count > 0) {
                parts.add(Texts.count(count) + " " + label);
            }
        });
        return Texts.count(kinds.size()) + " declarations (" + String.join(", ", parts)
                + "), not listed here — read one with `type`";
    }

    /**
     * The error declarations, whichever array Central filed them in.
     *
     * <p>HTTP-11. Filtering {@link TypeDef.ErrorDef} alone found 56 of {@code ballerina/http}'s 65 public error
     * declarations, and the nine it missed were disproportionately the LISTENER-side ones, published under an
     * alias category rather than under {@code errors}. The signal is {@code distinct}, and it is exact rather
     * than a heuristic: a type alias may be {@code distinct} only if it is an error type or an object type, and
     * an object type is an {@link TypeDef.ObjectDef} by the time it reaches here.
     */
    private static boolean isErrorDeclaration(TypeDef typeDef) {
        return typeDef instanceof TypeDef.ErrorDef
                || (typeDef instanceof TypeDef.Alias alias && alias.type().name().startsWith(DISTINCT));
    }

    private static List<TypeDef> errorsOf(Library library) {
        return library.declarations().stream().filter(Overview::isErrorDeclaration).toList();
    }

    // -----------------------------------------------------------------------
    // Worked code
    // -----------------------------------------------------------------------

    /**
     * The readme's Ballerina code, quoted whole. {@link Snippets} decides what counts as Ballerina.
     *
     * <p>Marked as a quotation with begin/end markers because it is one: these are the package author's bytes,
     * not this document's claim, and the oracle that checks every signature this view prints has to cut them out
     * by structure rather than by recognising a heading.
     */
    private static void quickstart(Report report, LoadedPackage loaded, String pkg) {
        List<String> blocks = Snippets.select(loaded);
        if (blocks.isEmpty()) {
            return;
        }
        report.heading(2, "Quickstart");
        // ADR-0013 asked for the rules a reader needs at the moment they read quoted code. ADR-0024 cut that to
        // two, because the third — what a mark on a line means — described a check that no longer runs.
        report.paragraph("*Every Ballerina block in the package's own readme, quoted verbatim and in its "
                + "order. It is Central's text and can be out of date; the signatures the container verbs "
                + "generate win wherever the two disagree.*");
        report.embedded(pkg + " readme usage", blocks.stream()
                .map(block -> "```ballerina\n" + block + "\n```")
                .collect(Collectors.joining("\n\n")));
    }

    /**
     * The guide's chunks, one line.
     *
     * <p>A chunk index rather than the chunks: this document is a map, and the map's job here is to say that a
     * recipe for the thing you want exists and what to type to read it.
     */
    private static void chunkIndex(Report report, LoadedPackage loaded, String pkg) {
        List<Guide.Chunk> chunks = Guide.chunksOf(loaded);
        if (chunks.isEmpty()) {
            return;
        }
        report.paragraph("Guide chunks (" + Texts.count(chunks.size()) + "): "
                + chunks.stream()
                        .map(chunk -> chunk.number() + ". " + Texts.code(chunk.title()))
                        .collect(Collectors.joining("  "))
                + " — " + Texts.code("bal library guide " + pkg + " <n>"));
    }

    // -----------------------------------------------------------------------
    // Next, and the rosters
    // -----------------------------------------------------------------------

    /**
     * Where to go, DERIVED from what the package declares rather than asserted.
     *
     * <p>ADR-0019: a pointer that cannot answer is worse than no pointer. The old bullet said {@code <path>}
     * unconditionally, and an agent following it on {@code ballerinax/aws.s3} spent a call to arrive back here.
     */
    private static List<String> nextBullets(Library library, String pkg) {
        List<String> next = new ArrayList<>();
        for (Surface.Scope scope : Surface.Scope.values()) {
            List<Surface.Container> containers = Surface.of(library, scope);
            if (containers.isEmpty()) {
                continue;
            }
            String argument = scope == Surface.Scope.MODULE || containers.size() > 1
                    ? ""
                    : " " + containers.get(0).name();
            next.add(Texts.code("bal library " + scope.verb() + " " + pkg + argument) + " — "
                    + prose(scope, containers));
        }
        next.add(Texts.code("bal library overview " + pkg + " -s \"<what you need>\"")
                + " — search every kind at once when you do not know which verb holds it");
        next.add(Texts.code("bal library type " + pkg + " <Name> [-r]")
                + " — a declaration whole, with the types it names");
        return next;
    }

    private static String prose(Surface.Scope scope, List<Surface.Container> containers) {
        return switch (scope) {
            case CLIENT -> containers.size() == 1
                    ? "the client's whole callable surface"
                    : Texts.count(containers.size()) + " clients, called with `->`";
            case CLASS -> Texts.count(containers.size()) + " classes and object types, called with `.`";
            case MODULE -> Texts.count(containers.get(0).functions().size())
                    + " functions callable without a client";
        };
    }

    private static void roster(
            Report report, String label, Surface.Scope scope, Library library, String pkg) {
        List<Surface.Container> containers = Surface.of(library, scope);
        if (containers.isEmpty()) {
            return;
        }
        if (scope == Surface.Scope.MODULE) {
            report.heading(2, label + " — " + Texts.count(containers.get(0).functions().size())
                    + ", call with " + Texts.code("."));
            report.literal(Report.columns(containers.get(0).standalone().stream()
                    .map(Fn.Standalone::name).toList()));
            return;
        }
        report.heading(2, label + " — " + Texts.count(containers.size()));
        List<Surface.Container> shown = containers.subList(0, Math.min(MAX_ROSTER_ROWS, containers.size()));
        report.bullets(shown.stream()
                .map(container -> Texts.code(container.name()) + " — " + counts(container) + " · "
                        + Texts.code("bal library " + scope.verb() + " " + pkg + " " + container.name()))
                .toList());
        if (shown.size() < containers.size()) {
            report.paragraph(Texts.count(containers.size() - shown.size()) + " more, not listed — "
                    + Texts.code("bal library " + scope.verb() + " " + pkg + " -s \"<what it does>\"")
                    + " searches all of them at once.");
        }
    }

    private static String counts(Surface.Container container) {
        List<String> parts = new ArrayList<>();
        int resources = container.operations().size();
        long remote = container.standalone().stream().filter(Fn.Remote.class::isInstance).count();
        long normal = container.standalone().stream().filter(Fn.Normal.class::isInstance).count();
        if (resources > 0) {
            parts.add(Texts.count(resources) + " resource");
        }
        if (remote > 0) {
            parts.add(Texts.count(remote) + " remote");
        }
        if (normal > 0) {
            parts.add(Texts.count(normal) + " normal");
        }
        return parts.isEmpty() ? "nothing callable" : String.join(", ", parts);
    }

    // -----------------------------------------------------------------------
    // The cross-kind search
    // -----------------------------------------------------------------------

    /** One thing a cross-kind search found: what it is, who owns it, and the command that opens it. */
    private record Hit(String kind, String label, String owner, String command) {

        String row() {
            return Texts.code(label) + " — " + kind + (owner.isEmpty() ? "" : " on " + Texts.code(owner))
                    + " · " + Texts.code(command);
        }
    }

    /**
     * {@code -s} across every kind at once.
     *
     * <p>This is the question no kind-specific verb can answer, because the caller does not yet know the kind —
     * and it is why {@code overview} takes the flag at all. Rows carry the kind, the owning symbol and the
     * runnable command, so one call turns "which symbol does X?" into an addressed follow-up.
     */
    private static String searched(LoadedPackage loaded, Options options) {
        String pkg = loaded.qualified().qualified();
        Library library = loaded.library();
        List<Hit> surface = new ArrayList<>();
        List<String> documented = new ArrayList<>();

        for (Surface.Scope scope : Surface.Scope.values()) {
            for (Surface.Container container : Surface.of(library, scope)) {
                Filter.Split<Fn> split = Filter.apply(options.search(), container.functions(),
                        Filter::surfaceOf, Filter::docsOf);
                split.surface().forEach(fn -> surface.add(callableHit(pkg, scope, container, fn)));
                split.documented().forEach(fn -> documented.add(nameOf(fn)));
            }
        }
        Filter.Split<TypeDef> types = Filter.apply(options.search(), library.declarations(),
                Filter::surfaceOf, Filter::docsOf);
        types.surface().forEach(typeDef -> surface.add(new Hit(kindOf(typeDef), typeDef.name(), "",
                "bal library type " + pkg + " " + typeDef.name())));
        types.documented().forEach(typeDef -> documented.add(typeDef.name()));

        List<Guide.Chunk> chunks = Guide.chunksOf(loaded).stream()
                .filter(chunk -> chunk.title().toLowerCase(Locale.ROOT)
                        .contains(options.search().toLowerCase(Locale.ROOT)))
                .toList();
        chunks.forEach(chunk -> surface.add(new Hit("guide chunk", chunk.title(), "",
                "bal library guide " + pkg + " " + chunk.number())));

        Report report = new Report("overview");
        report.heading(1, pkg + " " + loaded.version().text() + " — " + Texts.code(options.search()));

        List<Report.Fact> facts = new ArrayList<>(Report.warning(loaded.warning()));
        facts.add(new Report.Fact("Filter", Texts.code(options.search()) + " — "
                + Texts.count(surface.size()) + " by name or type, " + Texts.count(documented.size())
                + " more by documentation only"));
        facts.add(new Report.Fact("Scope", "every kind: clients, classes, module functions, declarations and "
                + "guide chunk titles"));
        report.facts(facts);

        report.heading(2, "Next");
        report.bullets(surface.isEmpty()
                ? List.of("read the map instead: " + Texts.code("bal library overview " + pkg),
                        "or widen the query — this one matched nothing, in any kind")
                : List.of("run the command on the row that looks right — each is complete as printed",
                        "add " + Texts.code("-r") + " to any of them for the types it names"));

        if (!surface.isEmpty()) {
            List<Hit> shown = surface.subList(0, Math.min(MAX_SEARCH_ROWS, surface.size()));
            report.heading(2, Texts.count(surface.size()) + " matched by name or type");
            report.bullets(shown.stream().map(Hit::row).toList());
            if (shown.size() < surface.size()) {
                report.paragraph(Texts.count(surface.size() - shown.size())
                        + " more, not listed. Narrow the query.");
            }
        }
        if (!documented.isEmpty()) {
            // NAMED, never rendered. Measured on github, `cache` is 6 surface hits against 33 in doc comments
            // alone; rendering those would bury the six the caller came for, and dropping them would lose the
            // vague query that has no lexical overlap with any identifier.
            report.heading(2, Texts.count(documented.size()) + " matched documentation only");
            report.literal(Report.columns(documented));
        }
        return report.toString();
    }

    private static Hit callableHit(String pkg, Surface.Scope scope, Surface.Container container, Fn fn) {
        String label = fn instanceof Fn.Resource resource
                ? resource.accessor() + " " + resource.paths().stream()
                        .map(io.ballerina.library.symbols.PathTree::displaySegment)
                        .collect(Collectors.joining("/"))
                : nameOf(fn);
        String argument = container.isModule() ? "" : " " + container.name();
        String selector = label.contains(" ") || label.contains("{") ? " '" + label + "'" : " " + label;
        return new Hit(kindOf(fn), label, container.isModule() ? "" : container.name(),
                "bal library " + scope.verb() + " " + pkg + argument + selector);
    }

    private static String nameOf(Fn fn) {
        return switch (fn) {
            case Fn.Standalone named -> named.name();
            case Fn.Constructor ignored -> "init";
            case Fn.Resource resource -> resource.accessor();
        };
    }

    private static String kindOf(Fn fn) {
        return switch (fn) {
            case Fn.Resource ignored -> "resource function, call with `->`";
            case Fn.Remote ignored -> "remote function, call with `->`";
            case Fn.Normal ignored -> "function, call with `.`";
            case Fn.Constructor ignored -> "constructor";
        };
    }

    private static String kindOf(TypeDef typeDef) {
        return switch (typeDef) {
            case TypeDef.Rec ignored -> "record";
            case TypeDef.Enumeration ignored -> "enum";
            case TypeDef.Alias ignored -> "type alias";
            case TypeDef.Constant ignored -> "constant";
            case TypeDef.Variable ignored -> "module-level variable";
            case TypeDef.ErrorDef ignored -> "error";
            case TypeDef.ObjectDef object -> object.form() == TypeDef.ObjectDef.Form.CLASS
                    ? "class" : "object type";
        };
    }
}
